### ISSUE TITLE
The function Pixbuf::new_from_vec

### ISSUE THREAD

**Issue by gbersac** [gtk-rs/gdk-pixbuf#13]
Trying to use that function, I had a really hard time using it and understanding [the function](https://github.com/gtk-rs/gdk-pixbuf/blob/master/src/pixbuf.rs#L34). What are the meaning of the arguments ? There is no documentation.

### `colorspace`
What is it ? For `new` it is `Color space for image`. I interpreted it as being the number of colour (red. green, blue) but this is supposed to be [n_channels](https://github.com/gtk-rs/gdk-pixbuf/blob/master/src/pixbuf.rs#L41). So I don't have a clue what it truly is.

I found this on the [gdk documentation website](https://developer.gnome.org/gdk-pixbuf/unstable/gdk-pixbuf-The-GdkPixbuf-Structure.html#GdkColorspace) :
```
The colorspace in which the samples are interpreted.
Flags: Read / Write / Construct Only
Default value: GDK_COLORSPACE_RGB
```

If I understand its value have to be `GDK_COLORSPACE_RGB`, is it useful to have this as argument ? If it is useful, why not create a `Colorspace` enum ?

### `bits_per_sample`
Since there is this line in the code :
```rust
assert!(bits_per_sample == 8);
```

There is only one correct value for `bits_per_sample`. Why not remove it from the argument of the list ?

### `row_stride`
I had to go on the [gdk documentation](https://developer.gnome.org/gdk-pixbuf/unstable/gdk-pixbuf-Image-Data-in-Memory.html#gdk-pixbuf-new-from-data) to understand what this argument is :

```
Distance in bytes between row starts
```

Should be in the doc !

### Conclusion
I'd agree to contribute but I don't know how the documentation of this crate is generated. Does regular doc comments works ?

**Comment by GuillaumeGomez** [comment]
On this crate yes (it's not generated like others). And sorry about this, for this release, we focused on providing as much docs as possible, as well as providing the best possible API. If you want to add documentation on this method, please do it, it'd be really appreciated! :)

**Comment by gbersac** [comment]
I'm ok to make a pull request but I have to make sure the comments I'll create will be correct (I'd rather have no comment than misleading one). What do you think about the comments I made ?

**Comment by GuillaumeGomez** [comment]
Don't worry, we'll help you through the PR. Your explanations are good, but if you could had some examples, it'd be very nice!

**Comment by gkoz** [comment]
@GuillaumeGomez Since we've moved most of the docs to the `lgpl-docs` repo wouldn't it be the applicable one?

**Comment by GuillaumeGomez** [comment]
I think it might be problematic to mix up gtk doc and non gtk doc and that's exactly what he did. Won't it be an issue if we start to change the doc file? Also: why this function hasn't doc generated? On the gnome website, there is [doc](https://developer.gnome.org/gdk-pixbuf/unstable/gdk-pixbuf-Image-Data-in-Memory.html#gdk-pixbuf-new).

**Comment by gkoz** [comment]
If there's interest in the `new_from_vec` function, it should be rewritten to take `Box<[u8]>` and named `new_from_data`. Under that name it will get documented by our tools (there is an entry for "new_from_data" but not for "new_from_vec").

I guess I agree we can drop `colorspace` and `bits_per_sample`.

**Comment by EdoardoLaGreca** [comment]
`let last_row_len = width * ((n_channels * bits_per_sample + 7) / 8);`
`assert!(vec.len() == ((height - 1) * row_stride + last_row_len) as usize);`
(src/pixbuf.rs lines 38 and 39)
Because of these lines my thread gets panicked, can someone please explain what do they do?

**Comment by GuillaumeGomez** [comment]
A picture can be seen as an array of arrays (each array would be a line of the image therefore). But since it's an image, you have more information. Generally you have 3 channels (colors) but if you have alpha in addition, then it's 4. I suppose there can be more channels of course. There is also the number of bits in each sample. I suppose you have all the information you need now.

**Comment by EdoardoLaGreca** [comment]
Not yet, what do you mean by "last row len"? What is the assert macro supposed to check by this condition?  
P.S. I don't know if it's supposed to be an integer but last_row_len will never be an integer because of "+ 7" and then "/ 8".

**Comment by braincow** [comment]
@EdoardoLaGreca I find myself in the same exact predicament right now. Did you ever manage to solve how to get it to work? My code:

```
        let pixbuf = Pixbuf::new_from_vec(
            image_buf,
            Colorspace::Rgb,
            false,
            8,
            600,
            508,
            3 * 508);
```

**Comment by braincow** [comment]
If anyone else stumbles here I solved this by using Image crate to originally parse the bitstream I get from reqwest. After that I can convert it into pixbuf from its bitstream. For most of you (and probably for me too) a redundant step to use Image in the first place, but this is how I got it to work: https://github.com/braincow/fmitestbed/blob/master/src/datapoint.rs

```
    pub fn image_as_pixbuf(&self) -> Pixbuf {
        let pixbuf = Pixbuf::new_from_vec(
            self.image.raw_pixels(),
            Colorspace::Rgb,
            false,
            8,
            self.image.width() as i32,
            self.image.height() as i32,
            3 * self.image.width() as i32);
        pixbuf
}
```
In above instead of image.raw_pixels() any Vec<u8> should work as long as you have the rest of the parameters correct (with alpha channel remember to multiply with four(4) to set row_stride correctly)

**Comment by GuillaumeGomez** [comment]
@braincow: Nice one! If you have enough motivation, writing an example showing this code in action would be really awesome!

**Comment by EdoardoLaGreca** [comment]
@braincow Can I use this function if my image is stored in a constant of type Vec\<u8\>? Which type should I use for your `image_as_pixbuf()` function? I am assuming you are using [this crate](https://crates.io/crates/image).

**Comment by braincow** [comment]
@EdoardoLaGreca yes, any Vec<u8> should work instead of image.raw_pixels(). For some reason though I did not get it to work directly from my original attempt (in the same source) with directly reading in the reqwests result, but Image read it in just fine and then I could dump it to pixbuf through that as shown. I think I might have had my image size originally incorrect or something.

Yes that is the Image crate I am using right now.

**Comment by braincow** [comment]
@GuillaumeGomez I am actually working on one for myself. :)

Being a Rust n00b however slows me down a bit. Currently I am having advanced lifetime issues in my code that I am having difficulties on getting around of (feel free to see the issue https://github.com/braincow/fmitestbed/issues/1).

If I ever get that app working I guess it would be a nice example overall on Gtk-rs (and adjacent) crates with couple of additional techniques and technologies showcased too.
